src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, Param, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+class Trie(map: Map[Char, Trie] = Map.empty,
+           textValue: Array[Char] = Array.empty,
+           normFunction: Char => Char = identity) extends Serializable {
+  val trieMap: Map[Char, Trie] = map
+  val value: Array[Char] = textValue
+  val normFunc: Char => Char = normFunction
+  def getValueString: String = this.value.toString
+
+  def getNodeValue(key: Array[Char], value: Array[Char]): Array[Char] = (key, value) match {
+    case (k, v) if k.length() == 1 => v
+    case (_, _) => Array.emptyCharArray
+  }
+  def put(key: String, value: String): Trie = this.put(key.toCharArray, value)
+
+  def put(key: Array[Char], value: String, isNormalized: Boolean = false): Trie = (key, value) match {
+    case (_, _) if key.length() == 0 => this.copy()
+    case (k, v) =>
+      val normalizedKey = if (isNormalized) k else k.map(this.normFunc)
+      val newLeaf = new Trie(textValue = getNodeValue(normalizedKey, v.toCharArray), normFunction = this.normFunc)
+      val next = this.trieMap.getOrElse(normalizedKey.head, newLeaf).put(normalizedKey.tail, v, true)
+      val newMap = this.trieMap + (normalizedKey(0) -> next)
+      new Trie(newMap, this.value, this.normFunc)
+  }
+
+  def get(letter: Char): Option[Trie] = this.trieMap.get(this.normFunc(letter))
+
+  def putAll(stringMap: Map[String, String]): Trie = {
+    stringMap.foldLeft(this){
+      case (out, (k, v)) => out.put(k.toCharArray, v)
```

Comment:
this can be `out.put(k, v)` right? Since you already have a `put` which takes `String`?

Reply:
I made the change, the initial idea was to have an external and internal contract. This was for convenience of the user of Trie.